### PR TITLE
feat: support `grouping` aggregate function

### PR DESCRIPTION
**Pull request by JasonLi-cn**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/5647

## Rationale for this change

Currently datafusion does not implement `grouping` function.
https://github.com/apache/datafusion/blob/4edbdd7d09d97f361748c086afbd7b3dda972f76/datafusion/physical-expr/src/aggregate/grouping.rs#L80-L84

## What changes are included in this PR?

Complete the `grouping` function.

https://www.postgresql.org/docs/9.5/functions-aggregate.html
https://learn.microsoft.com/en-us/sql/t-sql/functions/grouping-transact-sql?view=sql-server-ver15

## Are these changes tested?

Yes

## Are there any user-facing changes?

Yes. Perhaps we need to include in the documentation instructions for the `grouping` function.
https://arrow.apache.org/datafusion/user-guide/sql/aggregate_functions.html

### CODE REVIEW

**Comment by JasonLi-cn** [comment]
Related work:
https://github.com/apache/datafusion/issues/2477
https://github.com/apache/datafusion/pull/2486

**Review by waynexia — COMMENTED** [review]
Thanks @JasonLi-cn for this 👍 

I noticed this from PG's [document](https://www.postgresql.org/docs/9.5/functions-aggregate.html):

>. The arguments to the GROUPING operation are not actually evaluated, but they must match exactly expressions given in the GROUP BY clause of the associated query level.

Do we need to do some verifications to make sure the param of `GROUPING` matches `GROUP BY`? Also I see the implementation of `GroupingGroupsAccumulator` assumes the input expr is column, but `GROUP BY` doesn't have such a constrain.

**Comment by JasonLi-cn** [comment]
> Thanks @JasonLi-cn for this 👍
> 
> I noticed this from PG's [document](https://www.postgresql.org/docs/9.5/functions-aggregate.html):
> 
> > . The arguments to the GROUPING operation are not actually evaluated, but they must match exactly expressions given in the GROUP BY clause of the associated query level.
> 
> Do we need to do some verifications to make sure the param of `GROUPING` matches `GROUP BY`? Also I see the implementation of `GroupingGroupsAccumulator` assumes the input expr is column, but `GROUP BY` doesn't have such a constrain.

Thanks @waynexia for your suggestion. I agree with you and I will make improvements according to your suggestions.

File: datafusion/physical-expr/src/aggregate/grouping.rs [hunk 1]
```diff
@@ -96,8 +113,172 @@ impl PartialEq<dyn Any> for Grouping {
...
+    let mut mask = 0;
+    for (i, index) in indices.iter().rev().enumerate() {
+        if grouping_set[*index] {
+            mask |= 1 << i;
+        }
+    }
+    mask
+}
+
+impl GroupsAccumulator for GroupingGroupsAccumulator {
```
[the remark this answers is not in the document]
Comment (JasonLi-cn) [reply datafusion/physical-expr/src/aggregate/grouping.rs:210]: I agree with you. I also wanted to realize `Accumulator`. However, I find that `Accumulator` cannot be implemented based on the current definition. When calling `update_batch`, we need to know the information of the current grouping set, so we need to add a parameter to `update_batch`. This is a big change, I need to get community's advice.

[the remark this answers is not in the document]
Comment (waynexia) [reply datafusion/physical-expr/src/aggregate/grouping.rs:216]: I see. `Accumulator::update_batch` doesn't pass `group_indices: &[usize]` in. It does not make sense to me if we maintain one in another way for this special `grouping` expr only for implementing `Accumulator`. Do you have any insight @alamb ?

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/grouping.rs:216]: I agree having some special case simply for the `grouping` aggregate that forces changes on all other aggregates isn't ideal

> When calling update_batch, we need to know the information of the current grouping set, so we need to add a parameter to update_batch

After reading  https://www.postgresql.org/docs/9.5/functions-aggregate.html I see that `grouping` is basically a special case that only makes sense in the context of grouping set (it provides some context into the grouping set).

Given it is so special, I wonder if we could special case it somehow 🤔 

One thing maybe we could do is to add another signature?

```
trait `GroupsAccumulator`  {
...
    /// Called with the information with what grouping set this batch belongs to.
    /// The default implementaiton calls `Self::update_batch` and ignores the grouping_set
    fn update_grouping_batch(
        &mut self,
        _values: &[ArrayRef],
        group_indices: &[usize],
        opt_filter: Option<&arrow_array::BooleanArray>,
        total_num_groups: usize,
        grouping_set: &[bool],
    ) -> Result<()> {
      self.update_batch(_values, group_indices, opt_filter, total_num_groups)
    }
...
```


And then we could make it clear in the documentation that the agregator calls `update_group_batch` but that most implementations can just implement `update_batch`

**Review by alamb — COMMENTED** [review]
Thanks for this PR @JasonLi-cn  and the review @waynexia 

I was not familiar with the `grouping` function before. Fascinating. I left some thoughts

I apologize for the delay in review. I have been very bust lately

**Pull request closed without merging**